Improve radar data acquisition

amotl opened the pull request:
Dear @meteoDaniel,

this implements #138 in order to also acquire radar data from `/weather/radar` besides `/climate_environment/CDC/grids_germany/{5_minutes,hourly,daily}/radolan`. It is based upon your #140.

After the refactorings coming from #188 and #189, I took the chance to rework your contribution coming from #140 based on these changes. It might save you quite some amount of work when picking this up again.

With kind regards,
Andreas.

gutzbenj commented on wetterdienst/dwd/radar/index.py:
```diff
@@ -103,15 +104,28 @@ def _create_fileindex_radar(
         files_server, columns=[DWDMetaColumns.FILENAME.value], dtype="str"
     )
 
+    # Some directories have both "---bin" and "---bufr" files within the same directory,
```
Are both needed or would bin for all be fine?

amotl replied:
I don't know if it's needed. I just found @meteoDaniel started the `RadarDataType` enumeration und after getting into the details of data myself, I found e.g. https://opendata.dwd.de/weather/radar/sites/px/boo/ has **both** data types within the same directory. So, I figured the respective routine for filtering the outcome based on the obtained `radar_data_type` argument would be appropriate here.

I believe the PR in general just cares about **properly accessing and indexing** the different locations where radar data is stored for now. When it comes to supporting the actual **download**, `collect_radar_data` only handles RADOLAN through `_collect_radolan_data`, so the acquisition of alternative radar data beyond RADOLAN has not made it into the highlevel API `DWDRadarRequest` yet. There's more to come here and I will try to continue working on that.

meteoDaniel commented:
Dear @amotl thanks for working further on that issue. I had a lot of trouble the last two weeks. 
did you take into account the changed data provision of the sweep data? It is now divided into two sub directories of simple and polarimetric filtered data.

amotl commented:
Dear Daniel,

> Did you take into account the changed data provision of the sweep data? It is now divided into two sub directories of simple and polarimetric filtered data.

I just recognized it and reflected it within [test_radar_sites_fileindex_sweep_hdf5](https://github.com/earthobservations/wetterdienst/pull/190/files#diff-2cfde04105bff0202801f0db9253f427R101-R116).

We should properly take that into account by adding yet another argument to the machinery, right? Like
```
class RadarFilterType(Enum):
    """ Enumeration for different Radar Filter Types"""

    SIMPLE = "simple"
    POLARIMETRIC = "polarimetric"
```

With kind regards,
Andreas.

amotl commented on wetterdienst/dwd/radar/metadata.py:
```diff
@@ -0,0 +1,91 @@
+    WN_REFLECTIVITY = "wn"
+    RX_REFLECTIVITY = "rx"
+
+    # https://opendata.dwd.de/weather/radar/radolan/
+    RW_REFLECTIVITY = "rw"
+    RY_REFLECTIVITY = "ry"
+    SF_REFLECTIVITY = "sf"
```
@meteoDaniel: I've just added these. Did I give them appropriate names? Are RW, RY and SF also "reflectivity" parameters?

meteoDaniel replied:
No RW: Is a radar based hourly precipitation sum and SF is the same product for 24h sum

kmuehlbauer replied:
`RY` is kind of quality corrected reflectivity data (5 minutes interval).

meteoDaniel replied:
@kmuehlbauer  thanks for clarify. Only information available was that RSS-Feed entry: https://www.dwd.de/DE/leistungen/opendata/neuigkeiten/opendata_jul2018_03.html

amotl replied:
Thanks for your feedback on this.

So, I believe it's just fine to call them `RW_REFLECTIVITY`, `RY_REFLECTIVITY` and `SF_REFLECTIVITY`. Please unresolve conversation if you have any objections.

amotl commented on wetterdienst/dwd/radar/metadata.py:
```diff
@@ -0,0 +1,91 @@
+    SWEEP_VOL_VELOCITY_V = "sweep_vol_v"
+    SWEEP_VOL_VELOCITY_Z = "sweep_vol_z"
+
+    # https://opendata.dwd.de/climate_environment/CDC/grids_germany/daily/radolan/
+    # https://opendata.dwd.de/climate_environment/CDC/grids_germany/hourly/radolan/
+    # https://opendata.dwd.de/climate_environment/CDC/grids_germany/5_minutes/radolan/
+    RADOLAN_GRID = "radolan_grid"
```
JFYI: I renamed this to `RADOLAN_GRID` in order to address resources within `climate_environment/CDC/grids_germany/{daily,hourly,5_minutes}`. Do you believe this is appropriate? Maybe `RADOLAN_CDC` would be better?

meteoDaniel replied:
@amotl as I mentioned above I think we have to rename it. RADOLAN_CDC would be fine for me

amotl replied:
> It sounds weird to me to say `RADOLAN_GRID` because RADOLAN is always on GRID. `RADOLAN_CDC` would be fine for me.

Thanks, I will be happy to adjust the naming.

amotl replied:
I've just addressed the renaming within cf73cc8.

amotl commented:
Dear Daniel,

thanks for adding 810c7a1 in order to start bringing this to the top-level `DWDRadarRequest` API object.

With my latest commit 22447d2, the indexing will support all of:

- https://opendata.dwd.de/weather/radar/composit/
- https://opendata.dwd.de/weather/radar/radolan/
- https://opendata.dwd.de/weather/radar/sites/
- https://opendata.dwd.de/climate_environment/CDC/grids_germany/daily/radolan/
- https://opendata.dwd.de/climate_environment/CDC/grids_germany/hourly/radolan/
- https://opendata.dwd.de/climate_environment/CDC/grids_germany/5_minutes/radolan/

With kind regards,
Andreas.

amotl commented on example/radar/radar_sweep_hdf5.py:
```diff
@@ -0,0 +1,94 @@
+
+def radar_hdf5_example():
+
+    log.info("Acquiring radar sweep data in HDF5")
+    request = DWDRadarRequest(
+        radar_parameter=RadarParameter.SWEEP_VOL_PRECIPITATION_V,
+        date_times=RadarDate.LATEST.value,
+        radar_site=RadarSite.BOO,
+        radar_data_type=RadarDataType.HDF5,
+    )
```
This will not discriminate between **simple** and **polarimetric** yet.

meteoDaniel replied:
I think we can solve this on two ways: 
- One would be to introduce another flag/enum 
- Or (my favorite) to introduce new radar_parameters. 
SWEEP_VOL_PRECIPITATION_V_SIMPLE and SWEEP_VOL_PRECIPITATION_V_POLAR
And adding a note in the documentation.

kmuehlbauer replied:
@meteoDaniel I'm not sure I get this. Would it not be:

- SWEEP_VOL_V_SIMPLE and SWEEP_VOL_V_POLAR
- SWEEP_PRECIPITATION_V_SIMPLE and SWEEP_PRECIPITATION_V_POLAR

amotl replied:
Thanks. Let's do variant b) and extrapolate these options into the list of RadarParameters, as both of you suggested.

amotl replied:
I've just looked at this in more depth. The problem with the distinguation on the `RadarParameter` level would be that this difference only appears for the data in HDF5-format, right? The data in the main folder (e.g. [1]) does not distinguish between both variants.

So, while I definitively would have preferred the variant to encode this into `RadarParameter`, we might be better off to use just **another** parameter for this, which is only applied when selecting the `RadarDataType.HDF5` variant.

What do you think?

[1] https://opendata.dwd.de/weather/radar/sites/sweep_pcp_v/boo/

amotl replied:
With 99749ab, only these tests will fail:
```
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_precipitation_hdf5_yesterday FAILED
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_precipitation_hdf5_timerange FAILED
```

The reason for that is that **simple** and **polarimetric** is not taken into account yet. I will introduce a respective parameter within the next iteration.

amotl replied:
I've added a new argument `subset` to the machinery with 27a0b49.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,329 @@
+
+    # data = []
+
+    # FIXME: Implement indexing by timestamp when needed.
+    if date_times[0] != RadarDate.LATEST.value:
+        raise NotImplementedError(
+            "Acquisition of generic radar data only supports LATEST file for now"
+        )
+
+    file_index = create_fileindex_radar(
+        parameter=parameter,
+        radar_site=radar_site,
+        radar_data_type=radar_data_type,
+    )
+
+    # Find latest file.
+    filenames = file_index["FILENAME"].tolist()
+
+    try:
+        latest_file = list(filter(lambda x: "-latest-" in x, filenames))[0]
+    except IndexError:
+        latest_file = filenames[-1]
+
+    # Make up single-entry response.
+    # TODO: Discuss which datetime to use here.
+    # TODO: Maybe decode timestamp from file header, e.g. RX272355, RW272250, DX280005
+    now = datetime.now()
+    payload = _download_generic_data(latest_file)
+    response = [(now, payload)]
```
By now, this only implements acquisition of `-latest-`.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,329 @@
+
+    # data = []
+
+    # FIXME: Implement indexing by timestamp when needed.
+    if date_times[0] != RadarDate.LATEST.value:
+        raise NotImplementedError(
+            "Acquisition of generic radar data only supports LATEST file for now"
+        )
+
+    file_index = create_fileindex_radar(
+        parameter=parameter,
+        radar_site=radar_site,
+        radar_data_type=radar_data_type,
+    )
+
+    # Find latest file.
+    filenames = file_index["FILENAME"].tolist()
+
+    try:
+        latest_file = list(filter(lambda x: "-latest-" in x, filenames))[0]
+    except IndexError:
+        latest_file = filenames[-1]
+
+    # Make up single-entry response.
+    # TODO: Discuss which datetime to use here.
+    # TODO: Maybe decode timestamp from file header, e.g. RX272355, RW272250, DX280005
+    now = datetime.now()
+    payload = _download_generic_data(latest_file)
+    response = [(now, payload)]
```
By now, `_collect_generic_radar_data` only implements acquisition of `-latest-` files.

amotl replied:
> need to implement full indexing by timestamp for `/weather/radar/{composit,radolan,sites}`

5e9073d might solve it. At least, there is a test for **dx reflectivity** already.

kmuehlbauer commented:
@amotl I'm currently testing this, but have some questions:

- how to specifiy `sweep_pcp_X` data (single precipitation scan data, where  `X` can be `Z`/`V`)
- how to specify `sweep_vol_X` data (volume scan, where `X` can be `Z`/`V`)
- how to request all elevations from `sweep_vol` data (`00` to `09`)

I'm trying to find my way through the package, but would need some guidance.

meteoDaniel commented on wetterdienst/dwd/radar/index.py:
```diff
@@ -0,0 +1,221 @@
+    Returns:
+        indexing file path relative to climate observations path
+    """
+    if parameter == RadarParameter.RADOLAN_GRID:
```
Can we exchange this if/else conditionals to a dict based mapping?
E.g.:
PARAMETER_PATH_MAPPING[parameter].format(time_resolution=time_resolution.value, period_type=period_type.value)

Note: .format allows you to add more parameter than given in the string. That makes it a very nice feature in this context

amotl replied:
While I am not totally against this, I would like to keep it that way for now until we have resolved the HDF5 untangling problem outlined below. I believe there are too many conditions and some of them might require manual adjustments, so that it will not be that easy to gain improved conciseness from introducing another mapping here.

> Note: `.format()` allows you to add more parameter than given in the string. That makes it a very nice feature in this context.

While I believe I know what you are pointing out here, please let me also point out that it is not so easy to account for putting the path separator `/` at the appropriate places when the path to be computed has more fragments than others.

Please consider this snippet re. `RADAR_PARAMETERS_SITES` as an example:
```
parameter_path = f"weather/radar/sites/{parameter.value}/{radar_site.value}"
if parameter in RADAR_PARAMETERS_WITH_HDF5:
    if radar_data_type is None:
        raise ValueError("RadarDataType missing [hdf5 or binary]")
    elif radar_data_type is RadarDataType.HDF5:
        parameter_path = f"{parameter_path}/{radar_data_type.value}"
```

When using `RadarDataType.HDF5` here, the path will have to be extended by another fragment and this fragment has to be separated by another path seperator `/`. Folding this into the graceful behavior we might gain from `.format()` you have pointed out will not be sufficient.

I believe the function `build_path_to_parameter()` is straightforward right now and will easily leave room for adjusting to all edge cases we are still trying to tackle. I believe introducing another mapping here will get us more in the way of implementing all these edge cases in a clean and concise manner.

meteoDaniel reviewed (commented):
Thanks for this huge effort !

meteoDaniel commented:
> @amotl I'm currently testing this, but have some questions:
> 
> * how to specifiy `sweep_pcp_X` data (single precipitation scan data, where  `X` can be `Z`/`V`)
> * how to specify `sweep_vol_X` data (volume scan, where `X` can be `Z`/`V`)
> * how to request all elevations from `sweep_vol` data (`00` to `09`)
> 
> I'm trying to find my way through the package, but would need some guidance.

Dear @kmuehlbauer you can access data with:

```
    request = DWDRadarRequest(
        radar_parameter=RadarParameter.SWEEP_VOL_PRECIPITATION_V,
        date_times=RadarDate.LATEST.value,
        radar_site=RadarSite.BOO,
        radar_data_type=RadarDataType.HDF5,
    )

```
Afterwards you have to call `request.collect_data`

We are actually not covering the different elevations in `sweep_vol`. I think we have to introduce a new enum/parameter for such data.

amotl commented on wetterdienst/dwd/metadata/time_resolution.py:
```diff
@@ -17,6 +17,8 @@ class TimeResolution(Enum):
     DAILY = "daily"
     MONTHLY = "monthly"
     ANNUAL = "annual"
+    MINUTE_5 = "5_minutes"
```
I recognized this as well. Currently, we have defined `MINUTE_1`, `MINUTE_5`, `MINUTE_15` and `MINUTES_10`. Following this, the prefix `MINUTE_` currently wins in terms of usage count:
```
$ ag MINUTE_ | wc -l
      24
$ ag MINUTES_ | wc -l
      17
```

So, we might want to instead adapt `MINUTES_10` to `MINUTE_10` in order to align them to each other?

meteoDaniel replied:
Yes I think adapt MINUTES_10 to MINUTE_10 is fine.

amotl replied:
Adjusted with 4addbfa. Thanks again.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+        parameter: What type of radar data should be collected
+        date_times: list of datetime objects for which radar data shall be acquired
+        time_resolution: the time resolution for requested data, either hourly or daily
+        radar_site:      Site of the radar if parameter is one of RADAR_PARAMETERS_SITES
```
I can't spot it. Can you point this out more specifically?

meteoDaniel replied:
There is a tab in there . All other lines do not have a tab.

amotl replied:
Fixed with 530e9588. Thanks.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+        list of tuples of a datetime and the corresponding file in bytes
+    """
+
+    if parameter == RadarParameter.RADOLAN_GRID:
```
Let's call it `RADOLAN_CDC` then, right?

amotl replied:
I've address this with 2ca3920.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+        )
+
+    else:
+        return _collect_generic_radar_data(
```
Currently, the logic discriminates between `RADOLAN_CDC`-types of data located at `/climate_environment/CDC/grids_germany` and the other data located at `/weather/radar`, which I named `generic`. I will be happy to introduce another name for that if you have any suggestions.

meteoDaniel replied:
I think it is enough to say: _collect_radar_data. The presence of _collect_radolan_data in combination with RADOLAN_CDC will help contributers and user to understand where they have to go to find the origin of the code.

amotl replied:
Fixed with 0238a44d1.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+    """
+    Collect raw radar data for COMPOSITE and SITES.
+
+    Remark: As of now, only the LATEST file (like "raa00-dx_10132-latest-boo---bin")
```
No problem at all. It's just that it hasn't been implemented at the time of writing this comment. Removed.

amotl commented:
> How to obtain `sweep_` data?

Right, thanks for answering already, @meteoDaniel. See also https://github.com/earthobservations/wetterdienst/blob/db4fae200b5872ab7b31587a4913d7525d83abde/example/radar/radar_sweep_hdf5.py#L62-L67

kmuehlbauer commented:
> Right, thanks for answering already, @meteoDaniel. There's also an example program which outlines this

Thanks @amotl, I've found the example. What would I need to add for `date_times` (or another kwarg) to get the latest 10 files? I sense, that this might not be implemented yet.

amotl commented:
> What would I need to add for date_times (or another kwarg) to get the latest 10 files? I sense, that this might not be implemented yet.

That is true. No way to filter by "latest 10" for now.

I also believe this would be a way more convenient option than accessing by specific timestamps at all for this type of data, where there isn't really anything "historical" to access. Otherwise, one would have to figure out specific time ranges upfront, taking the respective intervals into consideration, etc. pepe.

kmuehlbauer commented:
> I also believe this would be a way more convenient option than accessing by specific timestamps at all for this type of data, where there isn't really anything "historical" to access.

True, the history is only three days IIRC. I would try to add code but I'm having a hard time to follow up through the abstraction levels.

amotl commented:
> I would try to add code but I'm having a hard time to follow up through the abstraction levels. 

No worries, let me do that for you. I am about to switch over to a more OO-based approach, essentially mitigating the tedious 
passing of arguments around.

Thanks for all of your suggestions.

amotl commented on docs/pages/api.rst:
```diff
@@ -245,31 +245,40 @@ Other variants::
+*****
 
-To use ``DWDRadolanRequest``, you have to provide a time resolution (either hourly or daily)
+To use ``DWDRadarRequest``, you have to provide a time resolution (either hourly or daily)
```
I've tried to rework the documentation. Please have a look again.

meteoDaniel commented:
> > What would I need to add for date_times (or another kwarg) to get the latest 10 files? I sense, that this might not be implemented yet.
> 
> That is true. No way to filter by "latest 10" for now.
> 
> I also believe this would be a way more convenient option than accessing by specific timestamps at all for this type of data, where there isn't really anything "historical" to access. Otherwise, one would have to figure out specific time ranges upfront, taking the respective intervals into consideration, etc. pepe.

But we still have a file_index of the files on the server whoch can help us to implement selecting of files from directory

amotl commented:
> But we still have a file_index of the files on the server which can help us to implement selecting of files from directory.

Sure, I will try to keep both options: a) Select files by timestamp(s) and b) select number of most recent files.

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+    data = []
+    # datetime = pd.to_datetime(datetime).replace(tzinfo=None)
+    for date_time in date_times:
+        if prefer_local:
```
Thanks for asking. As I am not that much into the specific details of `prefer_local` vs. `write_file` (and honestly I am not a big fan either), I didn't touch anything here.

You will recognize that I haven't implemented these options for `_collect_radar_data` at all, see https://github.com/earthobservations/wetterdienst/blob/4addbfaf33ebd24dff471c6578c53cfe929f8926/wetterdienst/dwd/radar/access.py#L89-L155

That way, the routine itself will become much more concise. Regarding the lacking functionality, we might consider shifting this to a different level than having it within the core functions.

amotl replied:
> That way, the routine itself will become much more concise.

So, I have deprecated the variant to store/restore to/from the local filesystem completely with 99749ab. However, the main function has been kept as `@deprecated`.

https://github.com/earthobservations/wetterdienst/blob/968fede64a0181692895a5a149643bca2516ad1c/wetterdienst/dwd/radar/access.py#L186-L250

amotl commented:
> Try to keep both options: a) Select files by timestamp(s) and b) select number of most recent files.

Saying this, I just found selecting files by timestamp will not be easy for all files in HDF5 format at all. You will see the problem when e.g. scanning the folder [1]:

- `00-20200926T14:50:57`
- `00-20200926T14:55:57`
- `01-20200926T14:51:20`
- `01-20200926T14:56:20`
- `02-20200926T14:51:43`
- `02-20200926T14:56:43`

Here, timestamps for the different elevations are not aligned with each other.

[1] https://opendata.dwd.de/weather/radar/sites/sweep_vol_v/boo/hdf5/filter_polarimetric/

amotl commented:
### HDF5 untangling problem
I just added 497dd60945fb, which denies accessing "latest" data for HDF5 again. There is no designation within the filename for this data (see also [1]) and we don't have logic yet to group by elevation level first and then compute the latest file per elevation level.

So, this adjustment should be correct in order to better be safe than sorry. Sorry that this is probably breaking `sweep_` HDF5-type acquisition completely now, because the way to address files by specific timestamps will also be not feasible, as outlined above.

[1] https://opendata.dwd.de/weather/radar/sites/sweep_vol_v/boo/hdf5/filter_polarimetric/

### Thoughts
When thinking about this again, I believe using the "latest X" acquisition option suggested by @kmuehlbauer within https://github.com/earthobservations/wetterdienst/pull/190#issuecomment-699967388 will be the only reasonable thing which might help us here.

amotl commented:
### `/radar/sites` timestamp untangling
Oh, when revisiting this again, I found that only data for `dx`, `pf`, `px`, `px250` and `sweep_pcp` within [1] have timestamps within its filenames aligned to straight intervals of 5 minute marks. Data within other folders deviate on the `minute` part.

[1] https://opendata.dwd.de/weather/radar/sites/

amotl commented:
After adding 99749ab and a18474da1019 (sigh!), only these tests will fail, which is perfectly fine.
```
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_pcp_v_hdf5_yesterday FAILED
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_pcp_v_hdf5_timerange FAILED
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_vol_v_hdf5_yesterday FAILED
tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_vol_v_hdf5_timerange FAILED
tests/dwd/radar/test_api_most_recent.py::test_radar_request_site_most_recent_sweep_pcp_v_hdf5 FAILED
tests/dwd/radar/test_api_most_recent.py::test_radar_request_site_most_recent_sweep_vol_v_hdf5 FAILED
```

The reason for that is that the **simple** vs. **polarimetric** subsets are not taken into account yet, see https://github.com/earthobservations/wetterdienst/pull/190#discussion_r495657973. I will introduce a respective parameter to address them properly within the next iteration.

amotl commented:
@kmuehlbauer: Now I am actually getting the "last 10 files" thing. 10 is not an arbitrary number here, but instead covers a whole set of `sweep_vol_v` data across all elevation levels (0-9) for one scan, right?

As outlined within https://github.com/earthobservations/wetterdienst/pull/190#issuecomment-700067511 and now implemented through a18474d, we will deny accessing any `RadarParameter.SWEEP_` data in HDF5 format using the `RadarDate.LATEST` option altogether. The rationale behind this is that there are no physical `-latest-` files on the DWD data repository and properly synthesizing them proved to be flaky for various reasons.

So, I've introduced the `RadarDate.MOST_RECENT` option, which should correctly implement [DWIM](https://en.wikipedia.org/wiki/DWIM) here.

After taking the **simple** vs. **polarimetric** subsets into account, the API to access `RadarParameter.SWEEP_` data in HDF5 format will be something like:
```
request = DWDRadarRequest(
    parameter=RadarParameter.SWEEP_VOL_VELOCITY_V,
    start_date=RadarDate.MOST_RECENT,
    site=RadarSite.BOO,
    format=RadarDataFormat.HDF5,
    subset=RadarDataSubset.POLARIMETRIC,
)
```

P.S.: I might remove the `radar_` prefix on all arguments to radar-specific methods and functions altogether as I believe this is redundant. Also, I might rename some other arguments. To make the long story short, I've amended the snippet above to use the new argument names yet to be implemented.

As always, I will be happy about any feedback to take into consideration.

amotl commented:
f235b8ff3a42 now adds the `subset` option as outlined within the last comment.

meteoDaniel commented:
Dear @amotl you can stash commits to reduce total amount of commits. E.g. git rebas -i HEAD~5 -> will display te last 5 commits. 
Change pick to s (=stash).

amotl commented:
> you can stash commits to reduce total amount of commits

Sure. We can also squash all commits just before merging.

amotl commented on wetterdienst/dwd/radar/sites.py:
```diff
@@ -0,0 +1,135 @@
+
+# https://docs.wradlib.org/en/stable/notebooks/radolan/radolan_network.html
+
+""" metadata about radar sites """
+RADAR_LOCATIONS = {
+    "asb": {
+        "site_name": "ASR Borkum",
+        "location_id": 10103,
+        "latitude": 53.564011,
+        "longitude": 6.748292,
+    },
+    "boo": {
+        "site_name": "Boostedt",
+        "location_id": 10132,
+        "latitude": 54.004381,
+        "longitude": 10.046899,
+    },
```
Dear @meteoDaniel and @kmuehlbauer,

[wradlib's RADOLAN Radar Network documentation](https://docs.wradlib.org/en/stable/notebooks/radolan/radolan_network.html) has a `get_radar_locations()` function. I just recognized that the lat/lon values are slightly different at some places and digits. Also, `get_radar_locations()` features the altitude of the station.

So, I believe just using the `get_radar_locations()` function here 1:1 will be a better option as I believe you verified the positions of these radar sites at least twice before writing them down, @kmuehlbauer?

With kind regards,
Andreas.

gutzbenj replied:
Only an alternative: We could parse the locations from here https://www.dwd.de/DE/derdwd/messnetz/atmosphaerenbeobachtung/_functions/HaeufigGesucht/koordinaten-radarverbund.pdf?__blob=publicationFile&v=5

kmuehlbauer replied:
@amotl In `wradlib-notebooks` I've taken them from a predecessor of the paper @gutzbenj linked above. There's also a note in that paper that some coords changed slightly. Need to fix this in `wradlib-notebooks` too.

This is only needed to plot the radar circles on top of radolan for visualisation.

amotl replied:
> Need to fix this in wradlib-notebooks too.

I see, [1].

[1] https://github.com/wradlib/wradlib-notebooks/blob/v1.8.0/notebooks/radolan/radolan_network.ipynb

amotl replied:
With dd294dbe819251, I've added a parser for [1] and updated the `RADAR_LOCATIONS` dictionary appropriately with its outcome.

Everyone can run this process on their own behalf by invoking
```
python -m wetterdienst.dwd.radar.sites
```

[1] https://www.dwd.de/DE/derdwd/messnetz/atmosphaerenbeobachtung/_functions/HaeufigGesucht/koordinaten-radarverbund.pdf?__blob=publicationFile

amotl commented on wetterdienst/dwd/radar/metadata.py:
```diff
@@ -0,0 +1,111 @@
+    #       seem to be present on opendata.dwd.de.
+    FX_REFLECTIVITY = "fx"
+    PG_REFLECTIVITY = "pg"
+    # PP_REFLECTIVITY = "pp"
```
Dear @meteoDaniel,

I found `PP_REFLECTIVITY` here but I haven't been able to find it neither at https://opendata.dwd.de/weather/radar/composit/ nor elsewhere on the public space of the DWD data repository.

Maybe it was just a typo? Please advise.

With kind regards,
Andreas.

meteoDaniel replied:
https://www.dwd.de/DE/leistungen/radarprodukte/radarkomposit_pp.pdf?__blob=publicationFile&v=2 

There are 4 composite types. I guess this document is outdated.

amotl replied:
Thanks. So, shall we just remove it from the code base?

amotl replied:
Removed through 204830b6b58.

amotl commented on wetterdienst/dwd/radar/metadata.py:
```diff
@@ -0,0 +1,111 @@
+
+
+RADAR_PARAMETERS_COMPOSITES = [
+    # RadarParameter.PP_REFLECTIVITY,
```
@meteoDaniel: You listed `PP_REFLECTIVITY` within `RADAR_PARAMETERS_COMPOSITES` the other day.

amotl commented:
Dear @kmuehlbauer and @meteoDaniel,

shall we also add acquisition for "radvor" data from [1]? I see that the data there has a suffix of `_000`-`_120`(in steps of 5) for "re" data [2] and `_000`, `_060`, `_120` for "rq" data [3]. Is this something we should account for when acquiring data?

[1] https://opendata.dwd.de/weather/radar/radvor/
[2] https://opendata.dwd.de/weather/radar/radvor/re/
[3] https://opendata.dwd.de/weather/radar/radvor/rq/

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,351 @@
+        # TODO: Discuss which datetime to use here.
+        # TODO: Maybe decode timestamp from file header,
+        #  e.g. RX272355, RW272250, DX280005.
+        now = datetime.now()
```
I am now using [`get_date_from_filename()`](https://github.com/earthobservations/wetterdienst/blob/radar-more/wetterdienst/dwd/radar/index.py#L53-L64) to derive the timestamp from the download file itself, see [`_download_generic_data(url: str)`](https://github.com/earthobservations/wetterdienst/blob/radar-more/wetterdienst/dwd/radar/access.py#L223-L259).

amotl commented:
Some of the `test_api_historic.py` tests apparently fail on Windows, see https://github.com/earthobservations/wetterdienst/pull/190/checks?check_run_id=1206522119.

Can I humbly ask one of you to run these on a Windows machine and report about the outcome?

```
pytest -vvvv -k test_api_historic
```

gutzbenj commented:
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_composite_historic_fx_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_pcp_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_vol_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached

gutzbenj commented:
Could it be realted to caching?

amotl commented:
> Could it be related to caching?

Hm. It seems to work flawlessly on Linux and macOS. The failures are coming from the transparent unzipping of .gz and .bz2 files introduced just recently.

You might want to test again after commenting out https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/access.py#L211.

amotl commented:
As I am currently clueless about this, I am humbly asking if you can investigate this further on your machine, @gutzbenj. What is noticeable is that all of the failing tests are all `_timerange`-like ones.

As processing the gzipped "radvor" data seems to work flawlessy also on Windows (check `pytest -vvvv -k test_radar_request_radvor`), it seems like this problem is revolving around all bzip2'ed artefacts processed with `bz2.BZ2File()`.

Throughout the DWD data repository, these .bz2 files can be found within:

- https://opendata.dwd.de/weather/radar/composit/fx/
- https://opendata.dwd.de/weather/radar/composit/wn/
- https://opendata.dwd.de/weather/radar/sites/sweep_pcp_v/boo/

amotl commented on wetterdienst/dwd/radar/metadata.py:
```diff
@@ -0,0 +1,127 @@
+    RY_REFLECTIVITY = "ry"
+    SF_REFLECTIVITY = "sf"
+
+    # /radvor
+    # https://opendata.dwd.de/weather/radar/radvor/
+    RE_REFLECTIVITY = "re"
+    RQ_REFLECTIVITY = "rq"
```
Dear @kmuehlbauer and @meteoDaniel,

I've also added support for acquisition of "radvor" data from [1].

I see that the data there has a suffix of `_000`-`_120`(in steps of 5) for "re" data [2] and `_000`, `_060`, `_120` for "rq" data [3]. Is this something we should account for when acquiring data?

Also, do you believe `RE_REFLECTIVITY` and `RQ_REFLECTIVITY` are appropriate enum names here?

With kind regards,
Andreas.

[1] https://opendata.dwd.de/weather/radar/radvor/
[2] https://opendata.dwd.de/weather/radar/radvor/re/
[3] https://opendata.dwd.de/weather/radar/radvor/rq/

meteoDaniel replied:
RADVOR is a forecast product and the steps you are mention are the lead time. I do not have any requirements to such a data. Nice feature.

kmuehlbauer replied:
@amotl This might be a nice feature for users wanting kind of forecast. I've not used it yet and can't say much about it.

amotl replied:
All right. Then just leave it like it is. `RE_REFLECTIVITY` and `RQ_REFLECTIVITY` are not obviously wrong names?

amotl commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,447 @@
+    return url, download_file_from_dwd(url)
+
+
+def _download_generic_data(url: str) -> Generator[RadarResult, None, None]:
+    """
+    Function (cached) that downloads radar data.
+
+    :param url:         The URL to the file on the DWD server
+
+    :return:            The file in binary, either an archive of one file
+                        or an archive of multiple files.
+    """
+
+    _, data = _download_generic_data_cached(url)
+
+    # RadarParameter.FX_REFLECTIVITY
+    if url.endswith(".tar.bz2"):
+        with bz2.BZ2File(data, mode="rb") as archive:
+            with tarfile.open(fileobj=archive) as tar_file:
+                for file in tar_file.getmembers():
+                    yield RadarResult(
+                        data=BytesIO(tar_file.extractfile(file).read()),
+                        timestamp=get_date_from_filename(file.name),
+                        filename=file.name,
+                    )
+
+    # RadarParameter.WN_REFLECTIVITY, RADAR_PARAMETERS_SWEEPS (BUFR)
+    elif url.endswith(".bz2"):
+        with bz2.BZ2File(data, mode="rb") as archive:
+            data = BytesIO(archive.read())
+            yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
+
+    # RADAR_PARAMETERS_RADVOR
+    elif url.endswith(".gz"):
+        with gzip.GzipFile(fileobj=data, mode="rb") as archive:
+            data = BytesIO(archive.read())
+            yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
+
+    else:
+        yield RadarResult(url=url, data=data, timestamp=get_date_from_filename(url))
```
After adding support for transparent unzipping of .gz and .bz2 files, some of the `test_api_historic.py` tests apparently fail on Windows with
```
EOFError: Compressed file ended before the end-of-stream marker was reached
```
See also https://github.com/earthobservations/wetterdienst/pull/190/checks?check_run_id=1206522119.

Can I humbly ask one of you to run these on a Windows machine and report about the outcome?

```
pytest -vvvv -k test_api_historic
```

amotl replied:
@gutzbenj has been able to confirm that:

```
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_composite_historic_fx_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_pcp_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
FAILED tests/dwd/radar/test_api_historic.py::test_radar_request_site_historic_sweep_vol_v_bufr_timerange - EOFError: Compressed file ended before the end-of-stream marker was reached
```

He asked: "Could it be related to caching?"

amotl replied:
It seems to work flawlessly on Linux and macOS. The failures are coming from the transparent unzipping of .gz and .bz2 files introduced just recently. Damn.

https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/access.py#L216-L252

> Could it be related to caching?

Hm. You might want to test again after commenting out https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/index.py#L45 and https://github.com/earthobservations/wetterdienst/blob/f576e4607d7088d5e4f76230d4a00218c0b63c72/wetterdienst/dwd/radar/access.py#L211.

amotl replied:
As I am currently clueless about this, I am humbly asking if you can investigate this further on your machine, @gutzbenj. What is noticeable is that all of the failing tests are all `_timerange`-like ones.
- test_radar_request_composite_historic_fx_timerange
- test_radar_request_site_historic_sweep_pcp_v_bufr_timerange
- test_radar_request_site_historic_sweep_vol_v_bufr_timerange

As processing the **gzipped** "radvor" data seems to work flawlessy also on Windows (check `pytest -vvvv -k test_radar_request_radvor`), it seems like this problem is revolving around all **bzip2**'ed artefacts processed with `bz2.BZ2File()`.

Throughout the DWD data repository, these .bz2 files can be found within:

- https://opendata.dwd.de/weather/radar/composit/fx/ (.tar.bz2)
- https://opendata.dwd.de/weather/radar/composit/wn/ (.bz2)
- https://opendata.dwd.de/weather/radar/sites/sweep_pcp_v/boo/ (.bz2)
  Same for `sweep_pcp_z`, `sweep_vol_v` and `sweep_vol_z`.

amotl replied:
4ae64c4 might have fixed it.

kmuehlbauer replied:
One more source might be flaky/not responding internet connection, when the data is processed before the download finished completely. :crossed_fingers: that this is resolved now.

amotl commented on docs/pages/api.rst:
```diff
@@ -252,31 +252,50 @@ Other variants::
+For ``RADOLAN_CDC``-data, the time resolution parameter (either hourly or daily)
+must be specified.
+
+The ``date_times`` (list of datetimes or strings) or a ``start_date``
+and ``end_date`` parameters can optionally be specified to obtain data
+from specific points in time.
```
`date_times` is no more. Adjust documentation.

meteoDaniel replied:
Are these notes for yourself @amotl ?

amotl commented on docs/pages/api.rst:
```diff
@@ -252,31 +252,50 @@ Other variants::
+from specific points in time.
+
+For ``RADOLAN_CDC``-data, datetimes are rounded to ``HH:50min``, as the
+data is packaged for this minute step. Additionally, you can provide a
+folder to store/restore RADOLAN data to/from the local filesystem.
```
> store/restore RADOLAN data to/from the local filesystem

This feature has not been carried over. Adjust documentation.

amotl commented on docs/pages/api.rst:
```diff
@@ -252,31 +252,50 @@ Other variants::
+
+This is an example on how to acquire ``RADOLAN_CDC`` data using
+``wetterdienst`` and process it using ``wradlib``.
+For a more thorough example, please have a look at `example/radar_radolan_cdc.py`_.
```
Link to whole `example/radar` folder here.

meteoDaniel commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -0,0 +1,447 @@
+    return url, download_file_from_dwd(url)
+
+
+def _download_generic_data(url: str) -> Generator[RadarResult, None, None]:
```
Why is url a `str` and not a `Path` object as we use in different other functions? 
Path has built in functions to check out suffix e.g.

amotl replied:
Using Path objects for operating on URLs hasn't always been without hassle in the past. So, we gradually shifted towards strings and URL operations using `urljoin()` instead of using Path objects at all. See also

https://github.com/earthobservations/wetterdienst/blob/643737573a1b867540b4079ebacd480bf850b933/wetterdienst/util/network.py#L9-L49

meteoDaniel commented on wetterdienst/dwd/radar/api.py:
```diff
@@ -0,0 +1,233 @@
+        if start_date == RadarDate.CURRENT and not self.period_type:
+            self.period_type = PeriodType.RECENT
+
+        # Evaluate "RadarDate.MOST_RECENT" for "start_date".
```
But can I ask for a single elevation at 15:03 and get the file for the 15:00 o clock scan?

kmuehlbauer replied:
I would not have dared to ask this, but this would be the icing of the cake.

amotl replied:
I believe `RadarDate.CURRENT` is what you are asking for here.

kmuehlbauer replied:
Not sure, if that's what @meteoDaniel wants, but would be nice to select only a specific sweep (of 00, 01, ..., 09) from the volume. As I said, I would not have dared to ask, since this could be added in a later PR (might get my hands dirty with that).

amotl replied:
Now I am getting it, you are aiming at a **single sweep** here. Hm, yeah. If this actually makes sense, I will see what I can do here.

kmuehlbauer replied:
A bit more context why users might want this, here is one example for using only a single sweep from the volume:

With higher elevation scans (>15 deg) one can process so called quasi vertical profiles (QVP). This is just a median over the azimuth for all ranges for this particular sweep. Stacking several sweeps along time it can be displayed like a time vs. height plot.

amotl replied:
This has been implemented with 7334c338.

meteoDaniel commented on wetterdienst/dwd/radar/index.py:
```diff
@@ -0,0 +1,251 @@
+    subset: Optional[RadarDataSubset] = None,
+    time_resolution: Optional[TimeResolution] = None,
+    period_type: Optional[PeriodType] = None,
+) -> str:
```
Maybe return Path objects here?

amotl replied:
Using Path objects for URLs has caused too much trouble in the past.

meteoDaniel replied:
Okay. Fine for me

meteoDaniel commented on wetterdienst/dwd/radar/store.py:
```diff
@@ -5,56 +5,64 @@
-    filepath = build_local_filepath_for_radolan(date_time, folder, time_resolution)
+    filepath = _build_local_filepath(parameter, date_time, folder, time_resolution)
 
     filepath.parent.mkdir(parents=True, exist_ok=True)
```
This will not work in case that the output of _build_local_filepath is a Union[str, Path].

amotl replied:
Thanks. `radar/store.py` is currently completely defunct.

meteoDaniel commented on wetterdienst/dwd/radar/store.py:
```diff
@@ -5,56 +5,64 @@
+    date_time: datetime,
+    folder: Union[str, Path],
+    time_resolution: TimeResolution,
 ) -> Union[str, Path]:
```
This

meteoDaniel commented on wetterdienst/dwd/radar/store.py:
```diff
@@ -5,56 +5,64 @@
 
-def build_local_filepath_for_radolan(
-    date_time: datetime, folder: Union[str, Path], time_resolution: TimeResolution
+def _build_local_filepath(
```
Output has to be Path, not Union[str, Path]

amotl replied:
Thanks. `radar/store.py` is currently completely defunct.

We will have to take care about that later, unless @gutzbenj might want to dedicate some time to that.

gutzbenj replied:
I'd suggest we remove it for now and add the feature in a later iteration.

amotl replied:
Removed with f4f5de6c6.

meteoDaniel commented on README.rst:
```diff
@@ -98,8 +101,9 @@ Get historical data for specific stations, using Python:
+    # Collect and analyse data here.
     for df in observations.collect_data():
-        # analyse the station here
+        pass
```
What should this loop do?

kmuehlbauer replied:
It would collect the data, I think. Maybe printing some output of the collected data would fit here.

amotl replied:
Thanks. I will say
```
# Collect and analyse data here.
for df in observations.collect_data():
    print(df)
```
here.

meteoDaniel commented on README.rst:
```diff
@@ -67,7 +70,7 @@ Details
 - Run SQL queries on the results.
 - Export results to databases and other data sinks.
-- Public Docker image on ghcr.io.
+- Public Docker image.
```
Nice ;)

gutzbenj commented on wetterdienst/dwd/radar/access.py:
```diff
@@ -225,6 +225,8 @@ def _download_generic_data(url: str) -> Generator[RadarResult, None, None]:
+    data.seek(0)
+
     # RadarParameter.FX_REFLECTIVITY
     if url.endswith(".tar.bz2"):
```
We may add this one to the enumeration I have created for all types of archive endings.

amotl replied:
Thanks. Implemented with b6897ce62.

kmuehlbauer commented:
This looks like it's almost finished. :tada: Is there any timeline when this will hit master and possibly be released? Thanks!

amotl commented:
Dear @kmuehlbauer,

> This looks like it's almost finished. 🎉

After some adjustments I made to this yesterday, I believe it **is** ready now. Please advise if I missed to address anything.

Cheers,
Andreas.

kmuehlbauer commented:
> After some adjustments I made to this yesterday, I believe it **is** ready now. Please advise if I missed to address anything.

OK, this is great news, @amotl! I can't immediately see any problems so far. If there's something amiss, it can still be added later.

amotl commented:
Finally, this has been merged. Thanks for all of your support along the way, @meteoDaniel, @gutzbenj and @kmuehlbauer!

> Is there any timeline when this will hit master and possibly be released? Thanks!

Will you cut a new release, @gutzbenj? Thanks already!
